accessor/cuda_helper.hpp

```diff
@@ -0,0 +1,200 @@
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_ACCESSOR_CUDA_HELPER_HPP_
+#define GKO_ACCESSOR_CUDA_HELPER_HPP_
+
+
+#include <type_traits>
+
+
+#if defined(__CUDACC__)
+#include <thrust/complex.h>
+#endif
+
+
+#include "block_col_major.hpp"
+#include "reduced_row_major.hpp"
+#include "row_major.hpp"
+#include "scaled_reduced_row_major.hpp"
+#include "utils.hpp"
+
+
+namespace gko {
+namespace acc {
+namespace detail {
+
+
+template <typename T>
+struct cuda_type {
+    using type = T;
+};
+
+// Unpack cv and reference / pointer qualifiers
+template <typename T>
+struct cuda_type<const T> {
+    using type = const typename cuda_type<T>::type;
+};
+
+template <typename T>
+struct cuda_type<volatile T> {
+    using type = volatile typename cuda_type<T>::type;
+};
+
+template <typename T>
+struct cuda_type<T*> {
+    using type = typename cuda_type<T>::type*;
+};
+
+template <typename T>
+struct cuda_type<T&> {
+    using type = typename cuda_type<T>::type&;
+};
+
+template <typename T>
+struct cuda_type<T&&> {
+    using type = typename cuda_type<T>::type&&;
+};
+
+
+// Transform std::complex to thrust::complex iff thrust::complex is included
+#if defined(__CUDACC__)
+template <typename T>
+struct cuda_type<std::complex<T>> {
+    using type = thrust::complex<T>;
+};
+#endif  // defined(__CUDACC__)
+
+
+}  // namespace detail
+
+
+/**
+ * This is an alias for CUDA's equivalent of `T`.
+ *
+ * @tparam T  a type
+ */
+template <typename T>
+using cuda_type_t = typename detail::cuda_type<T>::type;
+
+
+/**
+ * Reinterprets the passed in value as a CUDA type.
+ *
+ * @param val  the value to reinterpret
+ *
+ * @return `val` reinterpreted to CUDA type
+ */
+template <typename T>
+std::enable_if_t<std::is_pointer<T>::value || std::is_reference<T>::value,
+                 cuda_type_t<T>>
+as_cuda_type(T val)
+{
+    return reinterpret_cast<cuda_type_t<T>>(val);
+}
+
+
+/**
+ * @copydoc as_cuda_type()
+ */
+template <typename T>
+std::enable_if_t<!std::is_pointer<T>::value && !std::is_reference<T>::value,
+                 cuda_type_t<T>>
+as_cuda_type(T val)
+{
+    return *reinterpret_cast<cuda_type_t<T>*>(&val);
+}
+
+
+/**
+ * Changes the types and reinterprets the passed in range pointers as a CUDA
+ * types.
+ *
+ * @param r  the range which pointers need to be reinterpreted
+ *
+ * @return `r` with appropriate types and reinterpreted to CUDA pointers
+ */
+template <int dim, typename Type1, typename Type2>
+GKO_ACC_INLINE auto as_cuda_range(
```

Comment:
I prefer to leave it as it is, since this way, it is more descriptive if you use it in a non-Ginkgo environment. I will look if it is possible to use `as_cuda_range` inside `as_cuda_type`, so it is transparent inside Ginkgo.